azext_iot/central/models/template.py

```diff
@@ -15,30 +15,78 @@ def __init__(self, template: dict):
             self.name = template.get("displayName")
             self.interfaces = self._extract_interfaces(template)
             self.schema_names = self._extract_schema_names(self.interfaces)
+            self.components = self._extract_components(template)
+            if self.components:
+                self.component_schema_names = self._extract_schema_names(
+                    self.components
+                )
+
         except:
             raise CLIError("Could not parse iot central device template.")
 
-    def get_schema(self, name, interface_name=""):
-        # interface_name has been specified, do a pointed lookup
-        if interface_name:
-            interface = self.interfaces.get(interface_name, {})
-            return interface.get(name)
+    def get_schema(self, name, is_component=False, identifier="") -> dict:
+        entities = self.components if is_component else self.interfaces
+        if identifier:
+            # identifier specified, do a pointed lookup
+            entry = entities.get(identifier, {})
+            return entry.get(name)
 
-        # find first matching name in any interface
-        for interface in self.interfaces.values():
-            schema = interface.get(name)
+        # find first matching name in any component
+        for entry in entities.values():
+            schema = entry.get(name)
             if schema:
                 return schema
 
         # not found
         return None
 
+    def _extract_components(self, template: dict) -> dict:
+        try:
+            dcm = template.get("capabilityModel", {})
+            if dcm.get("contents"):
+                rootContents = dcm.get("contents", {})
+                components = [
+                    entity
+                    for entity in rootContents
+                    if entity.get("@type") == "Component"
+                ]
+
+                if components:
+                    return {
+                        component["name"]: self._extract_schemas(component)
+                        for component in components
+                    }
+
+        except Exception:
+            details = "Unable to extract schema for component from template '{}'.".format(
+                self.id
+            )
+            raise CLIError(details)
+
+    def _extract_root_interface_contents(self, dcm: dict):
+        rootContents = dcm.get("contents", {})
```

Comment:
hmm - we seem to do the same thing in extract components, can we extract getting the contents of a rootComponent into a method? Perhaps call it on construction of the Template?

Reply:
wont fix - for readability it is good to keep the calls as is.  call under _extract_components is for components where as call under _extract_root_interface_contents is to add under interfaces.